### DISCUSSION TITLE
docker hub latest and version specific tags

### DISCUSSION THREAD

**Discussion by justsomeguyontheinternet** [pymedusa/Medusa#9981] [Ideas & Feature requests]
The latest tag is common practice in docker, it is the tag used if a tag is not specified when deploying a container e.g. pymedusa/medusa. As a docker user I would like the latest tag to be the most recent stable release on an ongoing basis.

As a docker user I would like each github release to have a matching release on docker hub e.g. pymedusa/medusa:0.5.18. This helps docker users select a specific version to deploy in docker if there are known issues with the latest version.

This could be done via github release action or done manually.

**Comment by OmgImAlexis** [comment]
Ideally `latest` should be either dev (nightly) or the newest stable.

**Reply by justsomeguyontheinternet** [reply]
in most common practice latest is the most recent stable release. there is some annoyance that latest is a bit of a misnomer for the dev/nightly reason that you mention.

https://blog.container-solutions.com/docker-latest-confusion

**Reply by OmgImAlexis** [reply]
> in most common practice latest is the most recent stable release

Nope, I work with containers a LOT and you'll actually find common practise is the newest build being pushed is used. Meaning that ones that're only pushing tagged releases will use stable BUT almost any pushing nightlies will push them as latest. This is VERY common.

**Reply by justsomeguyontheinternet** [reply]
i disagree. i don't know if there's some distinction between a development local machine and the practice used on docker hub. i just checked my server where i run docker containers, 13 out of 13 deployed containers use the practice i described. there are other common examples including many upstream dependencies.

- pymedusa/medusa (up until 1 month ago)
- linuxserver all of their containers

> This image provides various versions that are available via tags. latest tag usually provides the latest stable version. Others are considered under development and caution must be exercised when using them.
- ubuntu

> The ubuntu:latest tag points to the "latest LTS", since that's the version recommended for general use. The ubuntu:rolling tag points to the latest release (regardless of LTS status).

> Along a similar vein, the ubuntu:devel tag is an alias for whichever release the "devel" suite on the mirrors currently points to, as determined by the following one-liner: wget -qO- http://archive.ubuntu.com/ubuntu/dists/devel/Release | awk -F ': ' '$1 == "Codename" { print $2; exit }'

- alpine (note edge)
> Supported tags and respective Dockerfile links:
20210804, edge
3.14.2, 3.14, 3, latest
- minio

> Run the following command to run the latest stable image of MinIO as a container using an ephemeral data volume:

> ```
> podman run \
>   -p 9000:9000 \
>   -p 9001:9001 \
>   minio/minio server /data --console-address ":9001"
> ```

- gitlab uses latest as i describe and has a separate nightly tag https://hub.docker.com/r/gitlab/gitlab-ce/tags
- nginx
- portainer
- emby

> latest: latest stable release
beta: latest beta release

**Reply by OmgImAlexis** [reply]
No offense but most of the examples you're using are either LTS which isn't what we follow here or they're by a group(linuxserverio) who honestly don't follow the norm, we'd had issues with this in the past.

Also "13 out of 13 deployed containers" isn't really that many to go by. I'm talking 1000s of daily contains. Most of which follow what I said...

**Reply by OmgImAlexis** [reply]
Also you have a brand new account.. Im a little more inclined to follow what im saying since I've been working on this for years vs some new random user dropping by.

**Reply by OmgImAlexis** [reply]
Checking my own systems apart from the linuxserverio containers not a single one follows what you're saying with nightlies pushed to their latest branch.

**Reply by justsomeguyontheinternet** [reply]
i'm not going to pretend that 13 container images is all the container images that exist, but it's a decent sample size and 100% of that sample is significant. like i said, maybe it's a distinction between development and docker hub deployments. i'm not sure what other reason there would be for having your users default to an unstable nightly image other than having more developers than users accessing the image repository. maybe 2 different channels on docker hub makes sense. pymedusa/medusa for users and pymedusa/medusadev can each have a latest following a different convention that makes more sense for that use case.

as another sample, all 25 images on the first page of docker hub use this convention. this includes images from docker inc. for docker and docker registry. i'm not going to check the other 8 million, but you're welcome to if you'd like.

you're welcome to believe whatever you want based on your own experience, but you don't need to be flippant and rude to someone who observed an issue, took the time to create a bug report, feature request, and provided relevant information. sorry this triggered you for whatever reason.

**Comment by medariox** [comment]
Like this? https://github.com/pymedusa/Medusa/pull/10008

**Reply by justsomeguyontheinternet** [reply]
thank you for doing that so quickly. it's definitely an improvement. "yes, but..." i think is the answer. let me explain by breaking down the feature request a little bit more.

- [x] must have - docker latest tag which is generally stable and kept up to date
- [ ] ideally - docker latest tag is built off of github release tags, not github master (since github master may differ from release tags)
- [ ] would be nice to have - version specific release tags are built and uploaded to docker hub.

the last nice to have would allow users to downgrade to a specific version that was known to be working if they had an issue with a newer release and wait for a fix/next release.

**Reply by p0psicles** [reply]
Maybe this could help? https://github.com/marketplace/actions/build-tag-publish-docker#tagging

**Reply by justsomeguyontheinternet** [reply]
there's a note at the top that link that says to use a newer repository, but yeah looks like that would help.

> PLEASE USE THE ORIGINAL REPO Since I modified it it got a bunch of new features, so I doubt you still need this.

newer links:
https://github.com/elgohr/Publish-Docker-Github-Action

https://github.com/marketplace/actions/publish-docker

looks like 2 similar options for tagging latest and specific release versions. the first i think aligns more with this project.
https://github.com/marketplace/actions/publish-docker#tags-1
https://github.com/marketplace/actions/publish-docker#tag_semver
